### PR TITLE
Improve Bloomfilter load and save speed by switching to a bit vector

### PR DESCRIPTION
**Pull request by gdoubleyew**
Improve Bloomfilter load and save speed by switching to a bit vector which can be read/written directly to disk. Load is 10x faster and save is 100x faster.

Task/Issue URL: 
Improve binary filter load time #6
https://github.com/duckduckgo/bloom_cpp/issues/6

Tech Design URL: 
CC: 

**Description**:
The approach to speed up load/save time is:

  - Change the vector to a generic bit vector where BlockType is char by default.
  - Add the logic to get/set a bit within a BlockType vector.
  - Add a file header to the saved Bloom Filter that includes version number, magic number, and various size parameters such as maxItems, sizeInBits etc. If the magic number isn't present then assume legacy file layout.
  - To load the vector from a file, get the vector.data() pointer and read directly into that memory address.
  - To write the vector to a file, get the vector.data() pointer and write it out directly.

I did some timing on this new implementation and compared to the previous.
Timing is with 3 million entries in bloomfilter, running on a Macbook Pro.

Lookup time:
new: 2.1 usec
old: 2.0 usec

Insert time:
new: 2.8 usec
old: 2.5 usec

Load time (cumulative all entries):
new: 0.21 sec
old: 2.68 sec

Save time (cumulative):
new: 0.03 sec
old: 2.27 sec
So load is 10x faster, save is 100x faster, lookup and insert 5-10% slower.

**Steps to test this PR**:
1. ./run_test.sh
1. speedtest.cpp to test speed of operations

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Comment by subsymbolic** [comment]
Many thanks for submitting this. Unfortunately we're all on other projects so it may take some time to review this but we will get to it. Thanks again for contributing!

File: src/BloomFilter.cpp [hunk 1]
```diff
@@ -18,8 +18,43 @@
...
+
+#define HDR_MAGIC1 0xBF
+#define HDR_MAGIC2 0xAA
+#define HDR_MAGIC3 0xFE
+#define HDR_MAGIC4 0xED
+
+#define VERSION_MAJOR 1
+#define VERSION_MINOR 0
+
+#define INLINE inline
```
Comment (subsymbolic) [inline src/BloomFilter.cpp:33]: Is this required? If not, let's remove it.

File: src/BloomFilter.cpp [hunk 2]
```diff
@@ -18,8 +18,43 @@
...
+
+
+// File Headers
+typedef struct LegacyHeader_s {
```
Comment (subsymbolic) [inline src/BloomFilter.cpp:37]: Legacy support isn't needed, our library and backend endpoint will be versioned. Let's keep the code simple and remove any legacy support or other references to the old bloom filter.

File: src/BloomFilter.cpp [hunk 3]
```diff
@@ -18,8 +18,43 @@
...
+    unsigned char e1;
+    unsigned char e2;
+    unsigned char e3;
+    unsigned char e4;
+} LegacyHeader;
+
+typedef struct FileHeader_s {
```
Comment (subsymbolic) [inline src/BloomFilter.cpp:44]: I really like this idea however it goes beyond the scope of speeding up the filter and needs some consideration as we would need to do extra work on our backend to support the new structure. I think we will probably do this in the future but let's consider that in a separate issue / PR.

File: test/BloomFilterTest.cpp [hunk 4]
```diff
@@ -16,19 +16,51 @@
...
+}
+
+static void setup() {
+    // Create the global bloomfilter data
+    BloomFilter bloomfilter1(FILTER_ELEMENT_COUNT, TARGET_ERROR_RATE);
+    gBloomData = createRandomStrings(FILTER_ELEMENT_COUNT);
+    gTestData = createRandomStrings(ADDITIONAL_TEST_DATA_ELEMENT_COUNT);
+    gTestData.insert(gBloomData.begin(), gBloomData.end());
+
+    // TODO use filesystem::temp_directory_path() to get platform independent tmp directory
```
Comment (subsymbolic) [inline test/BloomFilterTest.cpp:60]: Resolve todos across the PR: We only use todos as reminders during dev, by the time a feature hits prod the todos should be resolved.

File: src/BloomFilter.cpp [hunk 5]
```diff
@@ -30,36 +65,97 @@ static unsigned int sdbmHash(string text);
...
-static BlockType pack(const vector<bool> &filter, size_t block, size_t bits);
-
-static vector<bool> readVectorFromFile(const string &path);
-
-static vector<bool> readVectorFromStream(BinaryInputStream &in);
-
-static void unpackIntoVector(vector<bool> &bloomVector,
-                             size_t offset,
-                             size_t bitsInThisBlock,
-                             BinaryInputStream &in);
```
Comment (subsymbolic) [inline src/BloomFilter.cpp:44]: Glad to see all this go!

**Comment by subsymbolic** [comment]
@gdoubleyew, thanks again for your submissions here and apologies for the long silence, we have been on other projects. The good news is we are dusting this repo off and making improvements over the next few days. In terms of this PR, it's a lot bigger than expected, a complete rewrite, well beyond just speeding up the filter.

We can't accept such large-scale changes, however, if you want to package up a minimal footprint of just your data structure and speed changes (no legacy support, headers etc) before we pick this up internally on Monday, that would be great. If not we'll close this out then and continue internally. Thanks again, your ideas and benchmarking will be a huge help to future work either way.

**Comment by gdoubleyew** [comment]
Ok, thanks for your comments Mia. I probably can't get to this by Monday so I will leave it to your internal process. I can be available for any follow-up questions or actions. Thanks! Grant

**Pull request closed without merging**